### PR TITLE
Docker build system

### PR DESCRIPTION
**Pull request by augustg**
This branch was branched from terabyte's branch "feature/docker". If this is merged, that branch can be deleted.

The docker build system is a set of shell scripts and "Dockerfiles" designed to build packages for multiple platforms. Because this would normally require us to install multiple linux distributions, docker is used to provide the build environment.

Everything is contained in the new directory "build/". In it is also a README file.
Running build-all.sh currently produces:
* A source package
* An ITG2 arcade compatible machine revision (x86)
* An opensuse rpm package (x86_64)
* An ubuntu compatible debian package (x86_64)

Docker is really fast and very easy to work with. Hopefully, this eliminates the need for a chroot environment.

Here is the full readme. The directory layout refers to the new build/ directory.

> Directory layout
> ----------------
> 
> artifacts: Will contain artifacts from successful builds.
> 
> deb: Contains files necessary to build a .deb-package.
> 
> docker-images: Contains build scripts for all docker images used to build packages for different distributions.
> 
> rpm: Contains files necessary to build an .rpm-package.
> 
> All scripts in this directory except "docker-build.sh" should automatically produce an artifact with little to no input.
> 
> Building all artifacts
> ----------------------
> Running build-all.sh currently produces:
> * A source package
> * An ITG2 arcade compatible machine revision (x86)
> * An opensuse rpm package (x86_64)
> * An ubuntu compatible debian package (x86_64)
> 
> Instructions:
> 1. Make sure your OS architecture is x86_64.
> 2. Install docker.
> 3. Make sure the latest git tag is the release version.
>    Requirements:
>    * Must start with a digit
>    * Must NOT contain the following characters -
> 4. Have a private rsa key available for signing the machine revision. The openitg wiki contains intructions on how to make one.
> 5. As root, run:
>    # ./build-all.sh private.rsa

### CODE REVIEW

**Comment by terabyte** [comment]
My hero!  Thank you SO much for finishing this work - I'm sure if everything checks out we'll get it merged.  The probability I find time to work on my ITG upgradicab in my garage just increased tenfold.

Any ideas about *running* an arcade binary in a docker container?  The mobo in my boxxor unit died and I had to move to modern, commodity hardware.  I will need to try to make the pads / lights work through docker somehow...

**Comment by augustg** [comment]
Too bad about the mobo. I currently don't know of anyone who has the ability (and will) to test the arcade binary on the stock hardware/image.
It might be possible to run it inside docker. But i don't think it'll be pretty. At the very least you can make sure all the libraries get loaded properly when you run the openitg binary.
This article gives an idea of what is possible/needed: http://fabiorehm.com/blog/2014/09/11/running-gui-apps-with-docker/

**Comment by concubidated** [comment]
@terabyte I don't think running the binary inside docker is going to work. It is rather simple to run on a modern OS with modern hardware though (with machine revision binary). Put together a script to set it up automatically. 

https://github.com/concubidated/itgpro2/blob/master/itg-installer.sh

This is used on at least 6 machines in Norcal. 

@augustg Good job on the docker build image! This will be very helpful.

**Comment by spigwitmer** [comment]
this is beautiful

That being said though, bad things happened when building the opensuse package from a clean checkout of that branch:

```+ make clean
Making clean in src
make[1]: Entering directory '/usr/src/packages/BUILD/openitg-beta3/src'
test -z "openitg GtkModule.so" || rm -f openitg GtkModule.so
cd libresample && make clean
make[2]: Entering directory '/usr/src/packages/BUILD/openitg-beta3/src/libresample'
make[2]: *** No rule to make target 'clean'.  Stop.
make[2]: Leaving directory '/usr/src/packages/BUILD/openitg-beta3/src/libresample'
Makefile:3339: recipe for target 'clean_libresample' failed
make[1]: *** [clean_libresample] Error 2
make[1]: Leaving directory '/usr/src/packages/BUILD/openitg-beta3/src'
Makefile:386: recipe for target 'clean-recursive' failed
make: *** [clean-recursive] Error 1
error: Bad exit status from /var/tmp/rpm-tmp.LPPep0 (%build)```

It's sort of bizarre because configure reports using the static copy of the libresample source. I'll look into it a little further.

It did work like a charm with the machine revision though. Thanks a ton for building this out for us.

**Comment by spigwitmer** [comment]
with the configure.ac fix, build-all.sh runs through just fine, creating a signed machine revision, source tarball, 2 Ubuntu deb packages, and an OpenSUSE RPM. Awesome.

Because of the version policies of some of the distributions and how the tag's used in the build scripts, our own versioning would have to change; Debian/Ubuntu for example needs at least an explicit numeric major. @openitg/core should we just move to a sane versioning scheme?

**Comment by augustg** [comment]
Yes, I agree. 
I have a suggestion to get a discussion started. We merge all outstanding PRs, complete the VS2015 update and nothing else. Then we release it as 0.9. Possibly as a release candidate first.
OpenITG has stated a goal in the front-page readme (aims to be a drop in replacement...). And correct me if I'm wrong. Aren't we feature complete by that definition? If so, we should soon be at 1.0.
The reason I said 0.9 above and not 1.0 is because I'd like us to see if we can fix issue #80 first. If we can, 1.0 would be a "unifying release" where everyone can play under the same conditions with amazing polling rate. 
And getting something out there before 1.0 gives us time to find bugs.

**Comment by spigwitmer** [comment]
Yeah I say we go with a `major.minor[.patch]` scheme from here on, making the next release 0.9.

Last call from anyone in core....

**Comment by terabyte** [comment]
I am a huge fan of versioning schemes that include the exact git sha1 the version was built from.  I helped write this when I was at Palantir, for example:
https://github.com/kostasniktas/gradle-flexversion

Obviously, we aren't using gradle and we aren't going to use this exact code, but what I do like to see is a version that looks like this:

VER-N-gX

Where VER is something like "1.0" or "0.9" or "1.0-rc"
Where N is something like "git rev-list --count HEAD"
and where X is "git rev-parse --short=12 HEAD"

Other than that, I don't really care what VER is (0.9, 1.0, whatever you guys think is appropriate.  I think if we are a feature-complete drop-in replacement, then calling it 1.0 makes sense!)

**Comment by fent** [comment]
Whoa another past Palantirian that likes this game 😮

**Comment by terabyte** [comment]
Sure, there are lots because of the DDR machine we kept at Palantir for a while - that was actually my idea, and the machine still lives in my garage to this day =)  After I left Palantir I took it with me.

**Comment by fent** [comment]
Thanks for setting it up while you were there. Was fun playing whenever I visited the Shire.

**Comment by spigwitmer** [comment]
I'd like to keep the version uniform for each packaging medium, and RPMs [cannot have a hyphen in the version](http://www.rpm.org/max-rpm/ch-rpm-file-format.html) because of how the release ID is parsed. Several other projects get around this by having something like `1.0git<commit>`, though I'd be in favor of not including the commit hash if the source for the build is a canonical release at an exact tag. 2-3 numbers will be easier to remember for our users than a partial SHA1 with the same major/minor/patch version.

**Comment by DataBeaver** [comment]
It's common for distributions to tweak the version numbers of packages slightly to suite their style.  On Debian the proper way to express release candidate versions is to use a tilde: 1.0~rc1.  Tilde is defined as a special separator which compares as earlier than the version without a tilde, i.e. 1.0 comes after 1.0~anything.  A version like 1.0git1234567 would compare as coming after 1.0, which is not something you want for release candidates.

**Comment by augustg** [comment]
I'm sure everyone involved here have full time jobs. And since the number of hours currently invested into this project is limited we could decide not to use release candidates (or git hashes), since they complicate things. If we want release candidates we would probably have to write some code to convert version numbers if it's to be completely automated.
Instead we could release 0.9, if people find bugs we make a 0.9.1 etc. And the versioning complexity does not go further than that. This way the precious few hours we have are not spent trying to engineer a state of the art versioning scheme that fits all distros. 
Basically I'm trying to avoid an relatively unimportant problem that we don't really gain anything from solving.
Thoughts?

**Comment by terabyte** [comment]
My main goal is this:  Always knowing where a version came from.

When a user says "I am running version X", I want to know that means THIS EXACT GIT HASH.  If X contains the hash, that solves that nicely.  If we want "real releases" to be a bare number like "1.0", "2.0.1", etc, that is fine, but just checking out the source code and typing "make" MUST NOT produce a build that "thinks" it is "1.0" or "1.0.1" or "1.0.1rc" or "1.0.1-dirty" or whatever, with no hash, because that is useless.

**Comment by terabyte** [comment]
(but besides meeting that - whatever you guys think is best works for me - I agree it doesn't have to be perfect.  We can include the git hash by default, and build "real" releases with a special flag that causes them to be bare, or we can detect if you have a tag checked out, use that, otherwise use the sha1-based version, etc.)

**Comment by augustg** [comment]
With those requirements added we could keep it mostly the way it is. git tag --abbrev=0 is currently used to produce a version number (code in configure.ac).
If we are at a tag it'll output "0.9". If we're 4 commits ahead of the latest tag it'll output 0.9-4-hash.
All we would have to do is replace dash with plus for some (all?) distros. That's manageable.

**Comment by augustg** [comment]
The latest commit should fix the last remaining issues with version numbering. Dashes in the string produced by "git describe" must be replaced by plus signs to not cause problems with dpkg-deb and rpmbuild. Here are some examples of what the output files will look like:

**When we are on a tag (official release)**
- openitg-0.9.tar.xz (source tarball)
- ITG 2 OpenITG-0.9.itg
- openitg-0.9-1.suse42.1.x86_64.rpm
- openitg_0.9-1_amd64.deb

**When we are ahead of a tag (unofficial dev build)**
- openitg-0.9+3+g088dd12.tar.xz
- ITG 2 OpenITG-0.9+4+g8863a4e.itg
- openitg-0.9+3+g088dd12-1.suse42.1.x86_64.rpm
- openitg_0.9+3+g088dd12-1_amd64.deb

Is there anything left to fix before this can be merged?

**Comment by spigwitmer** [comment]
just did a successful build of everything after a merge from origin/master. we're good. Thanks again!
